packages/upload-client/src/fetch-with-upload-progress.js

```diff
@@ -1,8 +1,160 @@
+ */
+const withUploadProgress = (options) => {
+  const { onUploadProgress, body } = options
+
+  const rsp = new Response(body)
+  // @ts-expect-error web streams from node and web have different types
+  const source = iterateBodyWithProgress(rsp.body, onUploadProgress)
+  const stream = iterableToStream(source)
+  return {
+    ...options,
+    body: stream,
+  }
+}
+
+// HTTP/1.1 and browsers
+
+/* c8 ignore start */
+/**
+ *
+ * @param {string} url
+ * @param {import('./types.js').FetchOptions} init
+ */
+const fetchXhr = (url, { onUploadProgress, ...init }) => {
+  if (onUploadProgress) {
+    return /** @type {Promise<Response>} */ (
+      new Promise((resolve, reject) => {
+        const xhr = new XMLHttpRequest()
+        xhr.open(init.method || 'GET', url, true)
+        xhr.upload.addEventListener('progress', (e) =>
+          onUploadProgress({
+            total: e.total,
+            loaded: e.loaded,
+            lengthComputable: e.lengthComputable,
+            url,
+          })
+        )
+        xhr.upload.addEventListener('loadend', (e) =>
+          onUploadProgress({
+            total: e.total,
+            loaded: e.loaded,
+            lengthComputable: e.lengthComputable,
+            url,
+          })
+        )
+        xhr.onreadystatechange = () => {
+          if (xhr.readyState === 4) {
+            // @ts-expect-error doesn't have to match Response 100%
+            resolve({
+              status: xhr.status,
+              statusText: xhr.statusText,
+              body: xhr.response,
+              ok: ((xhr.status / 100) | 0) == 2,
+            })
+          }
+        }
+        xhr.onerror = (err) => reject(err)
+        Object.entries(init.headers || {}).forEach(([key, value]) =>
+          xhr.setRequestHeader(key, value)
+        )
+        /**
+         * @type {XMLHttpRequestBodyInit}
+         */
+        // @ts-expect-error ReadableStream as body is not supported by XHR
+        const body = init.body
+        xhr.send(body)
+      })
+    )
+  } else {
+    return fetch(url, init)
+  }
+}
+
+/* c8 ignore stop */
+
+// Deno supports H1 streaming
+const isNode =
+  typeof process !== 'undefined' &&
+  process.versions.node &&
+  !process.versions.bun &&
+  !process.versions.deno
+const isBrowser = typeof globalThis.XMLHttpRequest !== 'undefined'
 
 /**
  * @type {import('./types.js').FetchWithUploadProgress}
  */
-export const fetchWithUploadProgress = (url, init) => {
-  return ipfsUtilsFetch.fetch(url, init)
+export const fetchWithUploadProgress = (url, init = {}) => {
+  // https://developer.mozilla.org/en-US/docs/Web/API/PerformanceResourceTiming/nextHopProtocol
+  /**
+   * @type {string | undefined}
+   */
+  const protocol =
+    // @ts-expect-error nextHopProtocol is missing from types but is widely available
+    performance.getEntriesByType('resource')[0]?.nextHopProtocol
+  const preH2 = protocol !== 'h2' && protocol !== 'h2c' && protocol !== 'h3'
```

Comment:
Should this be the other way around - specify the protocols that do not support streaming?

Reply:
I could make it the other way around. the others are http/0.9, http/1.0 and http/1.1 I think

nextHopProtocol has a fixed set of values